### DISCUSSION TITLE
LaTeX Autocomplete

### DISCUSSION THREAD

**Discussion by nikkujensen** [sagemathinc/cocalc#6348] [Ideas]
Could autocomplete be a feature in the CoCalc LaTeX. I personally like [how TeXstudio accomplishes](https://tex.stackexchange.com/questions/125815/add-to-texstudio-autocompletion) this.

**Comment by williamstein** [comment]
I don’t claim it is very smart but right now ctrl+space does autocomplete. 

More will be added soonish, eg via LSP integration for useful context sensitive info.

**Reply by nikkujensen** [reply]
Ok, on mac is it command+space?

**Reply by williamstein** [reply]
No, because on Mac command+space is typically taken already by splotlight search, so it's not available for use in a web app.  You can also use option + space.  

<img width="444" alt="image" src="https://user-images.githubusercontent.com/1276278/212793273-53ebb48f-f1d0-4a67-8c5c-bd1c119b4eba.png">

**Reply by nikkujensen** [reply]
> More will be added soonish, eg via LSP integration for useful context-sensitive info.

@williamstein 
Will these additions include autocomplete for macro-like actions? For instance, I type `\begin{` and then use the autocomplete for `\begin{equation}`. Will there be functionality that automatically adds the `\end{equation}` too when selecting from the dropdown?

**Reply by williamstein** [reply]
>  Will there be functionality that automatically adds the \end{equation} too when selecting from the dropdown?

It does automatically do that already, but it happens when you hit enter after \begin{equation}, not immediately right when you select.  It happens even if you made \begin{equation} without using autocomplete.

Today I also made it so you can just use the tab key to do completions. If nothing is selected and there is something before tab it completes.

**Reply by nikkujensen** [reply]
Ahh, thank you. The tab is more natural for me.

**Reply by williamstein** [reply]
Do you have any wishlish items related to ChatGPT integration and latex?  I added a "help fix" button yesterday.  ChatGPT is also capable of doing many other clever things, e.g.,:

- you can write something out by hand, then send it a screenshot, and it turns that into latex
- it can generate a lot of content, e.g., diagrams, templates (e.g., a report or resume)
- it *can't* reliably help at all with bibliographies or references

**Reply by nikkujensen** [reply]
The main thing I can think of is probably a spin-off of "help fix". A wish is to streamline discovering how to do certain commands. I have spent lots of time searching things like: how to format a long latex document (i.e. a book or thesis), how to create macros commands for repeated processes, how to work the SVG to PDF workflow, finding the names of certain mathematical characters, etc. 

So an idea/wish is to use something like the "help fix" but instead find a new solution, not fixing a bug.

**Reply by haraldschilly** [reply]
For creating LaTeX, you can do this already. In the chat box you get an answer and you can even reply for further questions or tuning. e.g. here I'm in a new document and I ask with context "None":

> Show me how to make a table with 3 columns, where the first two are aligned left, and the last one aligned right. that table should have full width on the page.

That's in the "AI" dialog, without context. You could also select some latex code and tell it to use the selection as context.

I don't think these tools are at a point where they'll do this fully automatic for you. Feels like the error rate is so high, that you are thrown back more often than you wish. But this "conversational" style, where it tries to explain something along the output, is IMHO a great balance. I.e. at least I didn't had to think about looking up a table package and sift through documentation. Thankfully, it also didn't treat me like a newbie.

In any case:

## Question

![Screenshot from 2023-04-26 11-10-55](https://user-images.githubusercontent.com/207405/234529366-46ef90f7-4212-4bd8-bc58-1f18b1fc42d0.png)

## Answer

I copied the output into a suitable place and manually added the package include.

```tex
\begin{tabularx}{\linewidth}{@{} X l r @{}}
  \textbf{Column 1} & \textbf{Column 2} & \textbf{Column 3} \\
  This is text in column 1 & This is text in column 2 & \$4.50 \\
  More text in column 1 & More text in column 2 & \$12.00 \\
\end{tabularx}
```

![Screenshot from 2023-04-26 11-12-00](https://user-images.githubusercontent.com/207405/234529797-149adcf4-b470-437b-b350-e21d02792289.png)

**Comment by williamstein** [comment]
> So an idea/wish is to use something like the "help fix" but instead find a new solution, not fixing a bug.

Since that's already implemented, but you didn't realize it is implemented, do you have any suggestions for how to change the UI to make it more discoverable?  The "Help fix" feature is extremely easily discoverable, but "Find a new solution..." isn't.

Thanks for any ideas.
